src/nvim/quickfix.c

```diff
@@ -1230,9 +1048,308 @@ static void qf_new_list(qf_info_T *qi, char_u *qf_title)
   qi->qf_lists[qi->qf_curlist].qf_id = ++last_qf_id;
 }
 
-/*
- * Free a location list
- */
+/// Parse the error format matches in 'regmatch' and set the values in 'fields'.
+/// fmt_ptr contains the 'efm' format specifiers/prefixes that have a match.
+/// Returns QF_OK if all the matches are successfully parsed. On failure,
+/// returns QF_FAIL or QF_NOMEM.
+static int qf_parse_match(char_u *linebuf, size_t linelen, efm_T *fmt_ptr,
+                          regmatch_T *regmatch, qffields_T *fields,
+                          int qf_multiline, int qf_multiscan, char_u **tail)
+{
+  char_u *p;
+  char_u idx = fmt_ptr->prefix;
+  int i;
+  size_t len;
+
+  if ((idx == 'C' || idx == 'Z') && !qf_multiline) {
+    return QF_FAIL;
+  }
+  if (vim_strchr((char_u *)"EWI", idx) != NULL) {
+    fields->type = idx;
+  } else {
+    fields->type = 0;
+  }
+
+  // Extract error message data from matched line.
+  // We check for an actual submatch, because "\[" and "\]" in
+  // the 'errorformat' may cause the wrong submatch to be used.
+  if ((i = (int)fmt_ptr->addr[0]) > 0) {  // %f
+    if (regmatch->startp[i] == NULL || regmatch->endp[i] == NULL) {
+      return QF_FAIL;
+    }
+
+    // Expand ~/file and $HOME/file to full path.
+    char_u c = *regmatch->endp[i];
+    *regmatch->endp[i] = NUL;
+    expand_env(regmatch->startp[i], fields->namebuf, CMDBUFFSIZE);
+    *regmatch->endp[i] = c;
+
+    if (vim_strchr((char_u *)"OPQ", idx) != NULL
+        && !os_file_is_readable((char *)fields->namebuf)) {
+      return QF_FAIL;
+    }
+  }
+  if ((i = (int)fmt_ptr->addr[1]) > 0) {  // %n
+    if (regmatch->startp[i] == NULL) {
+      return QF_FAIL;
+    }
+    fields->enr = (int)atol((char *)regmatch->startp[i]);
+  }
+  if ((i = (int)fmt_ptr->addr[2]) > 0) {  // %l
+    if (regmatch->startp[i] == NULL) {
+      return QF_FAIL;
+    }
+    fields->lnum = atol((char *)regmatch->startp[i]);
+  }
+  if ((i = (int)fmt_ptr->addr[3]) > 0) {  // %c
+    if (regmatch->startp[i] == NULL) {
+      return QF_FAIL;
+    }
+    fields->col = (int)atol((char *)regmatch->startp[i]);
+  }
+  if ((i = (int)fmt_ptr->addr[4]) > 0) {  // %t
+    if (regmatch->startp[i] == NULL) {
+      return QF_FAIL;
+    }
+    fields->type = *regmatch->startp[i];
+  }
+  if (fmt_ptr->flags == '+' && !qf_multiscan) {  // %+
+    if (linelen >= fields->errmsglen) {
+      // linelen + null terminator
+      p = xrealloc(fields->errmsg, linelen + 1);
+      fields->errmsg = p;
+      fields->errmsglen = linelen + 1;
+    }
+    STRNCPY(fields->errmsg, linebuf, linelen);
```

Comment:
I kept 5 occurrences of `STRNCPY`/`STRNCPY`, so the lint job is expected to fail.